WIP - Add require-autocomplete lint rule

andrewmcgov opened the pull request:
## Description

<!--
Please include a summary of what you want to achieve in this pull request.

If this fixes an existing issue, please include the issue number.
-->

## Type of change

<!--
Remember to indicate the affected package(s), as well as the type of change for each package.
-->

- [ ] <!--Package Name--> Patch: Bug (non-breaking change which fixes an issue)
- [ ] <!--Package Name--> Minor: New feature (non-breaking change which adds functionality)
- [ ] <!--Package Name--> Major: Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [ ] I have added a changelog entry, prefixed by the type of change noted above (Documentation fix and Test update does not need a changelog as we do not publish a new version for these changes)

andrewmcgov commented on packages/eslint-plugin/lib/rules/react-require-autocomplete.js:
```diff
@@ -0,0 +1,137 @@
+  'url',
+  'week',
+];
+const autocompleteElementNames = ['input', 'TextField'];
```
If we have other components that some libraries may want the rule applied to, I _think_ we could make this a setting.

Tom-Bonnike replied:
Ha, I was curious about whether we had the `autocomplete-valid` rule and it seems like we apply it on the [<Select> component](https://github.com/Shopify/web-configs/blob/9852a937a6b51edc60215dcc34c1b943cb648fab/packages/eslint-plugin/lib/config/rules/jsx-a11y.js#L75-L80) too!

andrewmcgov commented on packages/eslint-plugin/lib/config/rules/shopify.js:
```diff
@@ -45,6 +45,8 @@ module.exports = {
   // Prefer all non-React-specific members be marked private in React class components.
   '@shopify/react-prefer-private-members': 'off',
+  // Require input elements to have autocomplete values
+  '@shopify/react-require-autocomplete': 'off',
```
Does this make the rule disabled by default, allowing apps to opt in on their own time?

BPScott replied:
Correct.

Tom-Bonnike commented on packages/eslint-plugin/tests/lib/rules/react-require-autocomplete.test.js:
```diff
@@ -0,0 +1,62 @@
+        return <TextField type="text" {...fieldProps} />;
+      }`,
+    },
+  ],
```
I think we could add two more test cases:
-  for an `input` type on which `autoComplete` is not allowed (e.g. `<input type="checkbox" />`)
- a test where it doesn’t warn about components other than `input` and `TextField`

Tom-Bonnike commented on packages/eslint-plugin/lib/config/react.js:
```diff
@@ -26,6 +26,7 @@ module.exports = {
       '@shopify/react-initialize-state': 'error',
       '@shopify/react-no-multiple-render-methods': 'error',
       '@shopify/react-prefer-private-members': 'error',
+      '@shopify/react-require-autocomplete': 'error',
```
I’m guessing you thought this through already but… could there be any case where we wouldn’t want an `autoComplete` on one of these input types? Asking because maybe it should simply be a `warn`.

Tom-Bonnike commented on packages/eslint-plugin/lib/rules/react-require-autocomplete.js:
```diff
@@ -0,0 +1,137 @@
+  }),
+};
+
+function findAttrValue(attributes, name, context) {
```
Interesting, I’d recommend using a module like [jsx-ast-utils](https://www.npmjs.com/package/jsx-ast-utils) so we don’t have to implement these utils ourselves?

Tom-Bonnike reviewed (commented):
Super cool, good work! Your commit still mentions it‘s a WIP you may want to amend that 😜

vsumner reviewed (approved):
Great work! Thank you for providing tests and the detailed documentation of this new rule.

I tested this against `Shopify/web` and because the rule is set to `error` for `react` it is indeed on by default.

There were 54 errors and they seem to be legitimate.

```
error: Input elements of type 'text' must have an autoComplete attribute (@shopify/react-require-autocomplete) at app/components/SkeletonIndex/SkeletonIndex.tsx:34:15:
  32 |             <SkeletonTabs />
  33 |             <Card.Section>
> 34 |               <TextField
     |               ^
  35 |                 connectedLeft={filterButton}
  36 |                 label=""
  37 |                 labelHidden


error: Input elements of type 'text' must have an autoComplete attribute (@shopify/react-require-autocomplete) at app/components/SsnField/SsnField.tsx:41:5:
  39 |
  40 |   return (
> 41 |     <TextField
     |     ^
  42 |       value={format(value)}
  43 |       label={label}
  44 |       helpText={helpText}
  ```

I think minor is safe since this is a new feature but not a breaking change.

andrewmcgov commented:
Thanks @vsumner and @Tom-Bonnike! 

I've update the PR with your feedback and added a changelog entry. 

- We're going to leave this as `error` by default and let individual repos turn it off if they want to. 
- Thanks to Tom's suggestion I used a couple helper functions from [jsx-ast-utils](https://www.npmjs.com/package/jsx-ast-utils). This lead to me being able to delete most of the code I had written for the rule 😅. I ran the same test that Victor ran on web and got 53 errors instead of 54, which is fine.
- I added a setting to the rule for repo's to add elements to have the rule applied to. For example to test in web I added the following to the `.eslintrc.js` file: 

```js
'@shopify/react-require-autocomplete': [
  'error',
  {
    inputComponents: ['TextField'],
  },
],
```

Let me know if I am missing anything else!

BPScott commented on packages/eslint-plugin/docs/rules/react-require-autocomplete.md:
```diff
@@ -0,0 +1,31 @@
+# Require inputs have have autocomplete attributes when necessary
+
+The HTML `autoComplete` helps users to complete filling in forms by using data stored in the browser. This is particularly useful for people with motor disabilities or cognitive impairment who may have difficulties filling out forms online.
```
```suggestion
The HTML `autocomplete` attribute helps users to complete filling in forms by using data stored in the browser. This is particularly useful for people with motor disabilities or cognitive impairment who may have difficulties filling out forms online.
```

andrewmcgov replied:
I'm using `autoComplete` with an uppercase `c` here because the rule is targeted towards React, and that the supported attribute has the uppercase letter: https://reactjs.org/docs/dom-elements.html#all-supported-html-attributes.

Do you think I should change the value here in the docs?

BPScott commented on packages/eslint-plugin/lib/rules/react-require-autocomplete.js:
```diff
@@ -0,0 +1,70 @@
+        return;
+      }
+
+      const hasAutoCompleteProp = hasProp(node.attributes, 'autoComplete');
```
Should attributes be lowercase rather than camel case? `autocomplete` rather than `autoComplete`?

BPScott commented on packages/eslint-plugin/lib/utilities/index.js:
```diff
@@ -147,6 +147,7 @@ module.exports = {
   polarisComponentFromJSX,
   getRootObject,
   docsUrl,
+  findDefinition,
```
Where is this used, can it be removed? It does not seem to be a top-level function in this file so iI'd be suprised if this even works.

BPScott reviewed (commented):
Code is mostly looking good. Added a few questions inline, mainly around capitalisation of attributes - i thought they'd always be lowercase.

BPScott commented on packages/eslint-plugin/CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@ and adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 <!-- ## Unreleased -->
```
Uncomment the heading, and add a subheading now there is changelog content for a new release

```suggestion
## Unreleased

### Changed
```

BPScott reviewed (approved):
Nice! Thanks for doing this, and thanks for teaching me `autoComplete` is one of those react uses different conventions edge-cases.

One nitpick about changelog formatting to fixup then you can go squash and merge!
